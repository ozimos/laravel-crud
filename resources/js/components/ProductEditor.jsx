import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ProductEditor extends Component {

  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit(event) {
    event.preventDefault();
    event.stopPropagation();
    if (this.props.canSubmit) {
      const { product, onAdd, onEdit, isEdit } = this.props;
      const submitMethod = isEdit ? onEdit : onAdd;
      submitMethod(product);
    }
  }

  render() {
    const { product, handleInput, handleChecked,
      canSubmit, isEdit } = this.props;
    const title = isEdit ? 'Edit' : 'Add New';
    return (
      <div>
        <h2>
          { `${title} Product`}
        </h2>
        <div>

          <form id="form" onSubmit={this.handleSubmit}>
            <div className="form-group">
              <label htmlFor="title">
Title
              </label>
              <input
                type="text"
                className="form-control"
                id="title"
                name="title"
                value={product.title || ''}
                aria-describedby="titleHelp"
                placeholder="Enter product title"
                onChange={handleInput}
                required
              />
              <small
                id="titleHelp"
                className="form-text text-muted"
              >
Enter a brief title for the product.
              </small>
            </div>
            <div className="form-group">
              <label htmlFor="description">Description</label>
              <textarea
                className="form-control"
                id="description"
                name="description"
                value={product.description || ''}
                placeholder="Description"
                onChange={handleInput}
                required
              />
            </div>
            <div className="form-group">
              <label htmlFor="price">Price</label>
              <input
                type="number"
                className="form-control"
                id="price"
                name="price"
                value={Number(product.price) || ''}
                aria-describedby="priceHelp"
                placeholder="Enter product price"
                onChange={handleInput}
                required
              />
            </div>
            <div className="form-group form-check">
              <input
                type="checkbox"
                className="form-check-input"
                id="availability"
                name="availability"
                checked={Boolean(product.availability) || false}
                onChange={handleChecked}
              />
              <label
                className="form-check-label"
                htmlFor="availability"
              >
Availabile

              </label>
            </div>
            <button
              type="submit"
              id="submit"
              className="btn btn-primary"
              disabled={!canSubmit}
            >
Submit

            </button>
          </form>

        </div>
      </div>);
  }
}

ProductEditor.defaultProps = {
  product: {},
  canSubmit: false,
  isEdit: false,
};

ProductEditor.propTypes = {
  onAdd: PropTypes.func.isRequired,
  onEdit: PropTypes.func.isRequired,
  handleInput: PropTypes.func.isRequired,
  handleChecked: PropTypes.func.isRequired,
  canSubmit: PropTypes.bool,
  isEdit: PropTypes.bool,
  product: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    price: PropTypes.oneOfType([
      PropTypes.string, PropTypes.number
    ]),
    availability: PropTypes.number,
  })
};
export default ProductEditor;
