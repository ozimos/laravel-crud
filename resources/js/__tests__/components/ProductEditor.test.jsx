/* global React:false shallow:false toJson:false */

import ProductEditor from '../../components/ProductEditor';

import { displayedProduct, newProduct }
  from '../__mocks__/productsMock';

describe('ProductEditor Component', () => {
  const props = {
    onAdd: jest.fn(),
    onEdit: jest.fn(),
    handleInput: jest.fn(),
    handleChecked: jest.fn(),
    canSubmit: true,
    isEdit: false,
    product: newProduct
  };
  const shallowSetup = () => shallow(<ProductEditor {...props} />);
  const event = {
    preventDefault: jest.fn(),
    stopPropagation: jest.fn()
  };
  it('should render without throwing an error', () => {

    const wrapper = shallowSetup();

    expect(wrapper.find('#form')
      .contains(
        <small
          id="titleHelp"
          className="form-text text-muted"
        >
          Enter a brief title for the product.
        </small>
      ))
      .toBe(true);
  });

  it('renders correctly', () => {
    let wrapper = shallowSetup();
    expect(toJson(wrapper)).toMatchSnapshot();

    const propsWhenEditing = {
      ...props,
      isEdit: true,
      product: displayedProduct };
    wrapper = shallow(<ProductEditor {...propsWhenEditing} />);
    expect(toJson(wrapper)).toMatchSnapshot();

    const propsWhenProductEmpty = {
      ...props,
      canSubmit: false,
      product: {} };
    wrapper = shallow(<ProductEditor {...propsWhenProductEmpty} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it(
    'should add a new product '
    + 'when the submit button on the add form is clicked',
    () => {
      const wrapper = shallowSetup();

      const handleSubmitSpy = jest.spyOn(
        wrapper.instance(),
        'handleSubmit'
      );
      wrapper.instance().handleSubmit(event);
      expect(handleSubmitSpy).toHaveBeenCalled();
    }
  );

  it(
    'should edit a  product '
    + 'when the submit button on the edit form is clicked',
    () => {
      const propsWhenEditing = {
        ...props,
        isEdit: true,
        product: displayedProduct };
      const wrapper = shallow(<ProductEditor {...propsWhenEditing} />);
      const handleSubmitSpy = jest.spyOn(
        wrapper.instance(),
        'handleSubmit'
      );
      wrapper.instance().handleSubmit(event);
      expect(handleSubmitSpy).toHaveBeenCalled();
    }
  );
});
