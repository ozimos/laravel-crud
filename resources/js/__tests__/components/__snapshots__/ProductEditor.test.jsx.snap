// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductEditor Component renders correctly 1`] = `
<div>
  <h2>
    Add New Product
  </h2>
  <div>
    <form
      id="form"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          aria-describedby="titleHelp"
          className="form-control"
          id="title"
          name="title"
          onChange={[MockFunction]}
          placeholder="Enter product title"
          required={true}
          type="text"
          value="estam"
        />
        <small
          className="form-text text-muted"
          id="titleHelp"
        >
          Enter a brief title for the product.
        </small>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="description"
        >
          Description
        </label>
        <textarea
          className="form-control"
          id="description"
          name="description"
          onChange={[MockFunction]}
          placeholder="Description"
          required={true}
          value="Dolor est eligendi sequi sapiente dicta eum.
   Et libero consequuntur laboriosam officia.
    Inventore velit esse possimus a voluptas mollitia."
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="price"
        >
          Price
        </label>
        <input
          aria-describedby="priceHelp"
          className="form-control"
          id="price"
          name="price"
          onChange={[MockFunction]}
          placeholder="Enter product price"
          required={true}
          type="number"
          value={73}
        />
      </div>
      <div
        className="form-group form-check"
      >
        <input
          checked={true}
          className="form-check-input"
          id="availability"
          name="availability"
          onChange={[MockFunction]}
          type="checkbox"
        />
        <label
          className="form-check-label"
          htmlFor="availability"
        >
          Availabile
        </label>
      </div>
      <button
        className="btn btn-primary"
        disabled={false}
        id="submit"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`ProductEditor Component renders correctly 2`] = `
<div>
  <h2>
    Edit Product
  </h2>
  <div>
    <form
      id="form"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          aria-describedby="titleHelp"
          className="form-control"
          id="title"
          name="title"
          onChange={[MockFunction]}
          placeholder="Enter product title"
          required={true}
          type="text"
          value="est"
        />
        <small
          className="form-text text-muted"
          id="titleHelp"
        >
          Enter a brief title for the product.
        </small>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="description"
        >
          Description
        </label>
        <textarea
          className="form-control"
          id="description"
          name="description"
          onChange={[MockFunction]}
          placeholder="Description"
          required={true}
          value="Dolor est eligendi sequi sapiente dicta eum.
   Et libero consequuntur non laboriosam officia.
    Inventore velit esse possimus a voluptas mollitia."
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="price"
        >
          Price
        </label>
        <input
          aria-describedby="priceHelp"
          className="form-control"
          id="price"
          name="price"
          onChange={[MockFunction]}
          placeholder="Enter product price"
          required={true}
          type="number"
          value={73}
        />
      </div>
      <div
        className="form-group form-check"
      >
        <input
          checked={true}
          className="form-check-input"
          id="availability"
          name="availability"
          onChange={[MockFunction]}
          type="checkbox"
        />
        <label
          className="form-check-label"
          htmlFor="availability"
        >
          Availabile
        </label>
      </div>
      <button
        className="btn btn-primary"
        disabled={false}
        id="submit"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`ProductEditor Component renders correctly 3`] = `
<div>
  <h2>
    Add New Product
  </h2>
  <div>
    <form
      id="form"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          aria-describedby="titleHelp"
          className="form-control"
          id="title"
          name="title"
          onChange={[MockFunction]}
          placeholder="Enter product title"
          required={true}
          type="text"
          value=""
        />
        <small
          className="form-text text-muted"
          id="titleHelp"
        >
          Enter a brief title for the product.
        </small>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="description"
        >
          Description
        </label>
        <textarea
          className="form-control"
          id="description"
          name="description"
          onChange={[MockFunction]}
          placeholder="Description"
          required={true}
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="price"
        >
          Price
        </label>
        <input
          aria-describedby="priceHelp"
          className="form-control"
          id="price"
          name="price"
          onChange={[MockFunction]}
          placeholder="Enter product price"
          required={true}
          type="number"
          value=""
        />
      </div>
      <div
        className="form-group form-check"
      >
        <input
          checked={false}
          className="form-check-input"
          id="availability"
          name="availability"
          onChange={[MockFunction]}
          type="checkbox"
        />
        <label
          className="form-check-label"
          htmlFor="availability"
        >
          Availabile
        </label>
      </div>
      <button
        className="btn btn-primary"
        disabled={true}
        id="submit"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
